### ISSUE TITLE
Does this not break verification search?

### ISSUE THREAD

**Issue by Kingdefender** [official-stockfish/Stockfish#1604]
Remove skipEarlyPruning argument to search() [https://github.com/official-stockfish/Stockfish/commit/8d1625d6df6d9918ca85cd3e169a06e3749c2f67](url)

I am probably hallucinating and nobody has questioned this, but I had some problems here. If you enter the verification search, you want to check for Zugzwang, so you want a move and check the situation after that. The original code goes immediately to the moves_loop to achieve just that. But in the verification search I don't think you can state anymore that `(ss-1)->currentMove == MOVE_NULL` to skip nullmove for instance. You have already called `pos.undo_null_move();`

Nobody has reported any problems. Marco is the author of skipEarlyPruning code but we just need some position checking perhaps. I have not checked Zugzwang positions yet. But I don't see how this patch does not break verification?  

A similar argument for IID. You want a move that you can use as ttMove. But if you enter null move and fail high, the null move is enough to fail high, you only can get a real move from the verification search in the new code. But that might be broken...

**Comment by Hanamuke** [comment]
It's possible that improvements can be made, but at the very least it's been tested therefore it doesn't have a measurable impact on playing strength. 

I'm not very knowledgable but we still have excludedMove, doesn't it do the job ?

**Comment by vondele** [comment]
@Kingdefender the statement '(ss-1)->currentMove != MOVE_NULL' is just there to avoid a null move after a previous null move. Proper verification search should be dealt with by the statement '(ss->ply >= thisThread->nmp_ply || ss->ply % 2 != thisThread->nmp_odd)'.

Have you seen any specific problems, or can you rephrase your issue more precisely ?

**Comment by Kingdefender** [comment]
Joost, thank you very much for the help. I only tried a few long term positions from the thread 'Some Zugzwang testpositions wanted for 'Enhanced Verification Search' (https://groups.google.com/d/msg/fishcooking/3di82Xke8nc/_CTOCkv0CQAJ) and everything so far seemed to work fine, with a little bit speed up in the deeper plies that I conclude must be from the more selectivity you mentioned.

So it almost could not be the verification search was not working and the only reason I could come up with is Günther Demetz' new code. I had not really figure out how exactly that worked, sorry about that,   that apparently it even could work  going a ply backwards again... If you can confirm that that is so. I'm much relieved. Maybe some comment in the text about this at some time. That I think leaves the Internal Iterative Deepening that was changed. Verification does not seem to be broken so that is good in case of nullmove, it can get a move from the verification search. And if it gets past nullmove and ProbCut without returning a fail high, in the new code I was wondering if now IID can be called recursively. If so, that might actually explain why your patch is so effective Joost :)

**Comment by vondele** [comment]
I've added a PR #1609 to try and clarify the current verification code.
I don't think I understand what you mean with 'it even could work going a ply backwards again'. 

IID can now be called iteratively, but on the other hand IID is worth about 2Elo only, so unlikely that small changes have big impact there.

**Comment by vondele** [comment]
@pb00068 thanks for checking. I think this issue can be closed.
